### PR TITLE
Shp

### PR DESCRIPTION
**Pull request by popkinj**
# Shapefile Upload
- Can now upload a shapefile to the map.
- Any defined projection is supported
- All files must be packaged in a shapefile

### CODE REVIEW

**Review by anissa-agahchen — DISMISSED** [review]

File: app/src/features/projects/components/ProjectLocationForm.tsx [hunk 1]
```diff
@@ -89,7 +91,44 @@ const ProjectLocationForm: React.FC<IProjectLocationFormProps> = (props) => {
...
+   * Convert a zipped shapefile to geojson
+   * @param e The file upload event
+   */
+  const handleShapefileUpload = (e: any) => {
+    // Only accept one file
+    const file = e.target.files[0];
+
+    // Back out if not a zipped file
+    if (file?.type !== 'application/zip') {
+      console.log('This does not look like a zip file.');
```
Comment (sdevalapurkar) [inline app/src/features/projects/components/ProjectLocationForm.tsx]: Lets remove this console log statement - later we can add our error dialog here but for now just the return will suffice

File: app/src/features/projects/components/ProjectLocationForm.tsx [hunk 2]
```diff
@@ -89,7 +91,44 @@ const ProjectLocationForm: React.FC<IProjectLocationFormProps> = (props) => {
     ]);
   }, [values.geometry]);
 
-  const handleSpatialUpload = async (e: any) => {
+  /**
+   * ## handleShapefileUpload
```
Comment (sdevalapurkar) [inline app/src/features/projects/components/ProjectLocationForm.tsx]:
```suggestion
```

File: app/src/features/projects/components/ProjectLocationForm.tsx [hunk 3]
```diff
@@ -89,7 +91,44 @@ const ProjectLocationForm: React.FC<IProjectLocationFormProps> = (props) => {
...
+      // The converter wants a buffer
+      const zip: Buffer = event?.target?.result as Buffer;
+
+      // Run the conversion
+      shp(zip).then((geojson) => {
+        const features = (geojson as any).features;
+        setFieldValue('geometry', [...features, ...values.geometry]);
+      }).catch((err) => {
+        const msg = 'Failed to convert shapefile to geojson'
+        console.error(msg,err);
```
Comment (sdevalapurkar) [inline app/src/features/projects/components/ProjectLocationForm.tsx]: remove log
```suggestion
```

File: app/src/features/projects/components/ProjectLocationForm.tsx [hunk 4]
```diff
@@ -165,10 +204,39 @@ const ProjectLocationForm: React.FC<IProjectLocationFormProps> = (props) => {
...
                 hidden
-                onChange={(e) => handleSpatialUpload(e)}
+                onChange={(e) => handleKMLUpload(e)}
               />
               Upload KML
             </Button>
+            <Tooltip
+              arrow
+              color='secondary'
+              title='Will only accept zipped shapefiles of a known projection.'
```
Comment (sdevalapurkar) [inline app/src/features/projects/components/ProjectLocationForm.tsx]: question:
can we specify what that known projection is perhaps??

File: app/src/features/projects/components/ProjectLocationForm.tsx [hunk 5]
```diff
@@ -165,10 +204,39 @@ const ProjectLocationForm: React.FC<IProjectLocationFormProps> = (props) => {
...
+                variant="outlined"
+                component="label"
+                size="medium"
+                color="primary"
+                disabled={isLoading}
+                style={{
+                  marginLeft:'1rem',
+                  border: '2px solid',
+                  textTransform: 'capitalize',
+                  fontWeight: 'bold'
```
Comment (sdevalapurkar) [inline app/src/features/projects/components/ProjectLocationForm.tsx]: do you mind extracting this out into a useStyles object so we can use the classname for both buttons? :)

**Comment by sdevalapurkar** [comment]
run `npm run format:fix` from `app` directory

File: app/src/features/projects/components/ProjectLocationForm.tsx [hunk 6]
```diff
@@ -89,7 +91,38 @@ const ProjectLocationForm: React.FC<IProjectLocationFormProps> = (props) => {
...
+    }
+
+    // Create a file reader to extract the binary data
+    const reader = new FileReader();
+    reader.readAsArrayBuffer(file);
+
+    // When the file is loaded run the conversion
+    reader.onload = (event: any) => {
+      // The converter wants a buffer
+      const zip: Buffer = event?.target?.result as Buffer;
```
Comment (NickPhura) [inline app/src/features/projects/components/ProjectLocationForm.tsx:119]: Do we need a null check on `zip` before called `shp` ?

File: app/package.json [hunk 7]
```diff
@@ -40,6 +40,7 @@
     "@tmcw/togeojson": "~4.2.0",
     "@turf/bbox": "~6.3.0",
     "@turf/boolean-equal": "~6.3.0",
+    "@types/shpjs": "^3.4.0",
```
Comment (NickPhura) [inline app/package.json]: This should move down to the dev dependencies section

File: app/src/features/projects/components/ProjectLocationForm.tsx [hunk 8]
```diff
@@ -89,7 +96,37 @@ const ProjectLocationForm: React.FC<IProjectLocationFormProps> = (props) => {
...
+
+      // Run the conversion
+      shp(zip).then((geojson: any) => {
+        const features = (geojson as any).features;
+        setFieldValue('geometry', [...features, ...values.geometry]);
+      });
```
Comment (NickPhura) [inline app/src/features/projects/components/ProjectLocationForm.tsx]: Can you change this to use async/await?

There could be some rare cases where regular .then style promises are useful/needed, but in general prefer async/await style.

**Review by sdevalapurkar — APPROVED** [review]

**Pull request merged**